File: java/src/main/java/org/revenj/server/commands/CRUD/Delete.java
package org.revenj.server.commands.CRUD;

import org.revenj.Utils;
import org.revenj.patterns.*;
import org.revenj.server.CommandResult;
import org.revenj.server.ServerCommand;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Optional;

public final class Delete implements ServerCommand {

	private final DomainModel domainModel;

	public Delete(DomainModel domainModel) {
		this.domainModel = domainModel;
	}

	public static final class Argument {
		public String Name;
		public String Uri;

		public Argument(String name, String uri) {
			this.Name = name;
			this.Uri = uri;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(ServiceLocator locator, Serialization<TInput> input, Serialization<TOutput> output, TInput data) {
		Argument arg;
		try {
			Type genericType = Utils.makeGenericType(Argument.class, data.getClass());
			arg = (Argument) input.deserialize(genericType, data);
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.Name);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Unable to find specified domain object: " + arg.Name);
		}
		PersistableRepository repository;
		try {
			repository = locator.resolve(PersistableRepository.class, manifest.get());
		} catch (ReflectiveOperationException e) {
			return CommandResult.badRequest("Error resolving repository for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		try {
			Optional<AggregateRoot> found = repository.find(arg.Uri);
			if (!found.isPresent()) {
				return CommandResult.badRequest("Can't find " + arg.Name + " with uri: " + arg.Uri);
			}
			repository.delete(found.get());
			return new CommandResult<>(output.serialize(found.get()), "Object deleted", 201);
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/commands/crud/Delete.scala
package net.revenj.server.commands.crud

import net.revenj.patterns._
import net.revenj.serialization.Serialization
import net.revenj.server.commands.crud.Delete.Argument
import net.revenj.server.{CommandResult, ServerCommand}

import scala.concurrent.Future

class Delete(domainModel: DomainModel) extends ServerCommand {

  override def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]] = {

    val arg = input.deserialize[Argument](data)
    lazy val manifest = domainModel.find(arg.get.Name)
    if (!arg.isSuccess) {
      CommandResult.badRequest(arg.failed)
    } else if (manifest.isEmpty) {
      CommandResult.badRequest(s"Unable to find specified domain object: ${arg.get.Name}")
    } else if (arg.get.Uri == null) {
      CommandResult.badRequest("Uri to find not specified.")
    } else if (!classOf[AggregateRoot].isAssignableFrom(manifest.get)) {
      CommandResult.badRequest(s"Specified type is not an aggregate root: ${arg.get.Name}")
    } else {
      val tryRepository = locator.resolve(classOf[PersistableRepository[AggregateRoot]], manifest.get)
      if (!tryRepository.isSuccess) {
        CommandResult.badRequest(s"Error resolving repository for: ${arg.get.Name}. Reason: ${tryRepository.failed.get.getMessage}")
      } else {
        import scala.concurrent.ExecutionContext.Implicits.global
        tryRepository.get.find(arg.get.Uri).flatMap {
          case Some(found) =>
            tryRepository.get.delete(found).map { _ =>
              val response = output.serializeRuntime(found)
              if (response.isSuccess) {
                CommandResult[TOutput](Some(response.get), "Object deleted", 200)
              } else {
                CommandResult[TOutput](None, response.failed.get.getMessage, 500)
              }
            }
          case _ =>
            CommandResult.badRequest(s"Can't find ${arg.get.Name} with uri: ${arg.get.Uri}")
        }
      }
    }
  }
}

object Delete {

  case class Argument(Name: String, Uri: String)

}